distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso

```diff
@@ -14,6 +14,7 @@ from Standard.Base.Data.Text.Encoding as Encoding_Module import Encoding, Encodi
 from Standard.Base.Error.Problem_Behavior as Problem_Behavior_Module import Problem_Behavior, Report_Warning
 import Standard.Base.Data.Locale
 import Standard.Base.Meta
+import Standard.Base.System.File.Existing_File_Behavior
```

Comment:
Importing something from `Standard.Base.System.File` seems to create a direct link from _core strata of the language_ (as `Text` is one of the core types) towards much less important _strata_ - e.g. I/O subsystem. I believe this goes against [principles of modularity](http://wiki.apidesign.org/wiki/NetBeans_Runtime_Container#Modularity_Manifesto). I believe the _language and core libraries_ shall be as small as possible, but _modularly extensible_ - be it Java or Enso.

This decay of mutual dependencies between different _stratas_ of StdLib is hard to prevent, when we don't have tools to ensure compile time failures when a _core_ strata attempts to call higher level _strata_. However that doesn't mean we should encourage such dependencies. Having reference to `File` in `Text.enso` is an example of such _strata crossing_ dependency and I'd suggest to avoid it.

I know where it leads. I've been there at least twice. With [NetBeans](http://wiki.apidesign.org/wiki/Modularization_of_NetBeans_Platform) and with [Java](http://wiki.apidesign.org/wiki/Modular_Java_SE). We end up with _(this time not object but) functionally oriented spaghetti  code_ and it is going to be hard to fix the mutual dependencies in future.

I guess the primary motivation for having `Text.write` is usability. Our users are more likely to find the method on existing `Text` object than some static function in some other library. Such a desire goes against the modularization effort. Huge problem when designing Java APIs, but other modern languages have an answer to that: [extension functions](https://kotlinlang.org/docs/extensions.html#extension-functions).

Doesn't Enso also supports extending other types with new functions/methods? Then the `write` function should be defined by `File.enso`, but register itself as an _extension_ into `Text` type. 

If possible, I'd find it way better design from modularity perspective.

Reply:
I'm kind of inclined to agree that we heading towards a spaghetti monster but at the same time PR implements what the ticket asked for

Reply:
I'm happy to move it to `File` if we decide that's the right approach.

Reg. extension functions - we have those yes, so I can rely on it. But I'm not sure if the language spec 1.0 includes them - IIRC they are removed so the situation may complicate a bit. Maybe we will be able to emulate the desired behaviour with a typeclass. 

So I can move the definition to `File.enso` - question is - depending on the language spec, if we may need to move it back here. But since this is a small operation, I think it is ok to do so.

Reply:
@JaroslavTulach That's a very good argument. Our stdlib is a little bit spaghetti because of lack of refactoring (which should be done in the meantime when implementing tasks) and also because of lack of proper tools to resolve it. It was one of the reasons why we are making changes to the language design. Currently, you don't have the tools in the language to resolve SOME of such a convoluted designs. In the past, we were thinking that we can introduce type classes to the language later in the future, but we realized that lack of them is very harmful right now - your comment shows one of such harmful behaviors The new lang design allows for making it nicer. We will fix the lang and then fix the libraries. On Monday the doc will be ready for reading, I'll ping you guys and I'd love to hear feedback about it from the perspective of this issue.